### ISSUE TITLE
Original php classes outside the cache folder are being edited in debug mode

### ISSUE THREAD

**Issue by xavier-cousin** [goaop/framework#191]
Hi,
We're using 0.4.3 and have limited the scope of aop to a single folder where our php classes actually using aop are (We're only using aop for dependency injection so ideally we only want the aop to act on these files and nothing else).

Here's the way we initialize the kernel:

```
    $debugMode = true;
    $applicationAspectKernel = ApplicationAspectKernel::getInstance();
    $applicationAspectKernel->init(array(
        'debug' => $debugMode,  // Use 'false' for production mode
        'appDir' => QSROOT.'/application/deps',
        // Cache directory
        'cacheDir' => QSROOT . '/cache',  // Adjust this path if needed
        // Include paths restricts the directories where aspects should be applied, or empty for all source files
        'includePaths' => array(
            QSROOT.'/application/deps',
        ),
        'interceptFunctions' => false
    ));
```

I'm pretty sure we misunderstand the way 'includePaths' and 'appDir' but we basically restrained both to that one folder I was talking about. Everything works fine when debug is false but when we set it to true, we suddenly see files being modified (the actual original php classes) when these files contain an include statement.
I can't provide any stack trace since there's no error thrown so I'm hoping you'll have an idea about what's going on just with this description.

Thanks!

**Comment by lisachenko** [comment]
Hello!

I faced this issue for several times and found a possible bug, when implementation of `realpath(false)` in PHP returns the current path, so framework can overwrite original files. This should be fixed in the next version, see https://github.com/lisachenko/go-aop-php/commit/b328977dc230f77b583479e3961a9f506c99fdb2.

Can you try to use 0.x-dev branch to check that this patch is working for you. If yes, I will make a release of 0.5.1

**Comment by lisachenko** [comment]
About config options:
- `appDir` should point to the top-level directory of your project. This path is used for calculation of relative paths in the `CachingTransformer`: 
  `$cacheUri    = str_replace($this->rootPath, $this->cachePath, $originalUri);` where `$this->rootPath` is equal to the `appDir`
- `cacheDir` is used in several places and determine the place where a transformed files, proxies, annotations, etc should be stored, this can be inside `appDir`, but add it into into the `excludePaths` option.
- `includePaths` option only affects weaving of AOP and represents a white-list of paths, where AOP should be applied, this should be directories under `appDir` or deeper.
- `excludePaths` is opposite to `includePaths` and allows to skip weaving for several files or directories. This can be useful to skip "bad" classes that can not be parsed statically, eg. spaghetti code.

However, be aware, that `includePaths` and `excludePaths` doesn't restrict the processing of all the code in your application, even if weaving is not required for some directories. This is needed for replacement of `include`|`require` and `new` expressions in the source code and for compatibility with opcode cachers in production mode.

**Comment by xavier-cousin** [comment]
I tried the 0.x-dev branch but it doesn't change the problem, the original files are still modified.
I guess I'm not using the appDir as expected since the path points to the one folder I want aop to be processed in. The bug probably comes from that fact originally but I'm not sure why.

What I would ideally want is being able to have aop look only at a specific path for files to process and leave anything else untouched and loaded though composer autoloader.

Do you think that's something achievable with this library?

**Comment by lisachenko** [comment]
@xavier-cousin I should investigate your issue more carefully. 

PS. To make an aop working only for specific path, just put it into `includePaths` option. This should be enough. But I'm not sure about pointing `appDir` to the sub-directory, this can lead to a strange results, because some files outside this directory can contain extra `include`s - they will be not processed by weaver.

**Comment by xavier-cousin** [comment]
I tried to have things configured as they should, using the includePaths to specify the paths where classes using aop are sitting in but for reasons we haven't completely identified yet, when we do that we get errors from many places in our legacy code with classes redeclared.
We don't have time to dive in it yet so for now the solution we came up with was to fork and make the aop class loader only focus on files within our "aop folder" and ignore anything else entirely. We know that in our case it won't be a problem because no file in our codebase includes directly these files so nothing needs rewriting.
We'll probably spend some time soon to understand why our legacy code was crashing and at that point I will come back to you if I find anything.
Thanks for your help!
